crates/oxc_linter/src/rules/react/jsx_key.rs

```diff
@@ -0,0 +1,395 @@
+            "#,
+            None,
+        ),
+        (
+            r#"
+            // testrule.jsx
+            const trackLink = () => {};
+            const getAnalyticsUiElement = () => {};
+
+            const onTextButtonClick = (e, item) => trackLink([, getAnalyticsUiElement(item), item.name], e);
+            "#,
+            None,
+        ),
+        (
+            r#"
+            function Component({ allRatings }) {
+                return (
+                  <RatingDetailsStyles>
+                    {Object.entries(allRatings)?.map(([key, value], index) => {
+                      const rate = value?.split(/(?=[%, /])/);
+
+                      if (!rate) return null;
+
+                      return (
+                        <li key={`${entertainment.tmdbId}${index}`}>
+                          <img src={`/assets/rating/${key}.png`} />
+                          <span className="rating-details--rate">{rate?.[0]}</span>
+                          <span className="rating-details--rate-suffix">{rate?.[1]}</span>
+                        </li>
+                      );
+                    })}
+                  </RatingDetailsStyles>
+                );
+              }
+              "#,
+            None,
+        ),
+        (
+            r#"
+            const baz = foo?.bar?.()?.[1] ?? 'qux';
+
+            qux()?.map()
+
+            const directiveRanges = comments?.map(tryParseTSDirective)
+            "#,
+            None,
+        ),
+        (
+            r#"
+            import { observable } from "mobx";
+
+            export interface ClusterFrameInfo {
+              frameId: number;
+              processId: number;
+            }
+
+            export const clusterFrameMap = observable.map<string, ClusterFrameInfo>();
+
+          "#,
+            None,
+        ),
+    ];
+
+    let fail = vec![
+        (r#"[<App />];"#, None),
+        (r#"[<App {...key} />];"#, None),
+        (r#"[<App key={0}/>, <App />];"#, None),
+        (r#"[1, 2 ,3].map(function(x) { return <App /> });"#, None),
+        (r#"[1, 2 ,3].map(x => <App />);"#, None),
+        (r#"[1, 2 ,3].map(x => x && <App x={x} />);"#, None),
+        (r#"[1, 2 ,3].map(x => x ? <App x={x} key="1" /> : <OtherApp x={x} />);"#, None),
+        (r#"[1, 2 ,3].map(x => x ? <App x={x} /> : <OtherApp x={x} key="2" />);"#, None),
+        (r#"[1, 2 ,3].map(x => { return <App /> });"#, None),
+        (r#"Array.from([1, 2 ,3], function(x) { return <App /> });"#, None),
+        (r#"Array.from([1, 2 ,3], (x => { return <App /> }));"#, None),
+        (r#"Array.from([1, 2 ,3], (x => <App />));"#, None),
+        (r#"[1, 2, 3]?.map(x => <BabelEslintApp />)"#, None),
+        (r#"[1, 2, 3]?.map(x => <TypescriptEslintApp />)"#, None),
+        (r#"[1, 2, 3]?.map(x => <><OxcCompilerHello /></>)"#, None),
+        ("[1, 2, 3].map(x => <>{x}</>);", None),
+        ("[<></>];", None),
+        //     (r#"[<App {...obj} key="keyAfterSpread" />];"#, None),
```

Comment:
Maybe an inline comment on why is this commented out.

Reply:
added diagnostic for it

Reply:
`KeyPropMustBePlacedBeforeSpread`